### PR TITLE
feat(hardhat): implement contract deployment persistence

### LINKED ISSUE #669: feat: Add persistence to Hardhat node and deployments
## Current Behavior
Currently, when the Hardhat container is stopped and restarted:
- All deployed contracts are lost
- The blockchain state is reset
- Deployment files are not persisted between container restarts

## Proposed Solution
We need to implement persistence for both the blockchain state and contract deployments:

### Short-term solution
1. Configure Docker volumes to persist deployment files:
   - Map `/app/deployments` to a named volume
   - Map `/app/artifacts` to a named volume
   - Map `/app/cache` to a named volume

### Long-term solution
Consider implementing one of these alternatives for full blockchain state persistence:

1. **Ganache Integration**
   - Replace Hardhat Network with Ganache
   - Benefits:
     - Built-in state persistence
     - Compatible with most Ethereum development tools
     - Easy to configure

2. **Full Ethereum Client**
   - Implement Geth or Hyperledger Besu
   - Benefits:
     - Production-like environment
     - Full blockchain persistence
     - More realistic testing conditions

Note: We attempted to implement block persistence using Hardhat's Ignition feature, but it proved unfeasible due to limitations in the Hardhat Network implementation. This led to the consideration of alternative solutions like Ganache or full Ethereum clients (Geth/Besu) for proper blockchain state persistence.

## Implementation Details
1. Update `docker-compose.yml` and 'Dockerfile.hardhat`
2. Add deploy function
3. Change hardhat config

## Additional Considerations
- Need to evaluate performance impact of persistent volumes
- Consider backup strategies for persistent data
- Document the process for clearing persistent data when needed
- Consider migration path from Hardhat Network to chosen alternative solution

## Related Issues
- None

## Breaking Changes
- None for the short-term solution
- Migration effort required for long-term solutions

/label enhancement
/label blockchain
/label persistence

### PR DESCRIPTION
**Pull request by mpaya5**
## Description
This PR implements persistence for contract deployments in the Hardhat development environment. It ensures that deployed contracts' information persists between container restarts by properly configuring Docker volumes and implementing manual deployment file saving.

## Changes
- Added volume mappings in docker-compose.yml for:
  - /app/deployments
  - /app/artifacts
  - /app/cache
- Updated Dockerfile.hardhat with proper permissions
- Implemented manual deployment file saving in deploy scripts
- Updated Hardhat configuration for correct deployment paths

## Testing Done
- Verified deployment file persistence after container restart
- Tested contract deployment process
- Verified file permissions and access
- Checked deployment file integrity

## Notes
This PR addresses only deployment persistence. Block/state persistence will require a separate solution (likely migration to Ganache or a full Ethereum client) as Hardhat Network doesn't support state persistence.

## Breaking Changes
None. This is a backwards-compatible enhancement.

## Related Issues
- Closes #669
- Depends on #626

## Checklist
- [x] Code follows project style guidelines
- [x] Changes are tested and working as expected
- [x] Documentation has been updated
- [x] Docker configuration has been validated
- [x] All tests pass
- [x] No new warnings introduced

/label enhancement
/label blockchain
/label persistence

### CODE REVIEW

File: docker/Dockerfile.hardhat [hunk 1]
```diff
@@ -11,18 +11,22 @@ RUN npm install --ignore-scripts
...
+    chmod -R 777 /app/cache
 
 ENV PATH="/app/node_modules/.bin:${PATH}"
 
-RUN echo -e '#!/bin/sh\necho "Compiling contracts..."\nnpx hardhat compile --force\necho "Starting Hardhat node..."\nexec ./node_modules/.bin/hardhat node --network hardhat' > /app/start.sh && \
-    chmod +x /app/start.sh && \
-    ls -la /app/start.sh
+# Copy the start script from hardhat/scripts instead of creating it
+COPY ./hardhat/scripts/start.sh /app/start.sh
+RUN chmod +x /app/start.sh
```
Comment (cristiam86) [inline docker/Dockerfile.hardhat:28]: @mpaya5 in the past we have had many problems executing shell scripts when the studio runs on Windows. Is there a way to test this on your side?

**Review by cristiam86 — APPROVED** [review]
Nice!
Please note that we must also deploy all the contracts and store their addresses somewhere so that they can be send to the front end.

**Comment by mpaya5** [comment]
image
@cristiam86 I created an issue related about that: https://github.com/yeagerai/genlayer-studio/issues/659

File: docker/Dockerfile.hardhat (hunk 1 above)
Comment (mpaya5) [reply docker/Dockerfile.hardhat]: You were right. I pushed a new commit solving that: https://github.com/yeagerai/genlayer-studio/pull/670/commits/a9ab9afe02b270827beab19ba36fb77da4c9ad59

Comment (AgustinRamiroDiaz) [reply docker/Dockerfile.hardhat]: @mpaya5 what was the issue? I'm curious

Comment (mpaya5) [reply docker/Dockerfile.hardhat]: The error was  that it couldn’t find the script.sh file on Windows. I’m not sure why this happened. This occurred after running `docker compose up.`

**Comment by mpaya5** [comment]
Closing this PR as the persistence functionality will be implemented as part of https://github.com/yeagerai/genlayer-studio/pull/713, which provides a more comprehensive solution including:

Contract deployment persistence through Docker volumes
Proper file handling and permissions
Integration with the consensus service
Complete deployment workflow
The requirements from this PR will be fully covered there as part of the larger consensus service integration. Please refer to https://github.com/yeagerai/genlayer-studio/pull/713 for the implementation.

**Pull request closed without merging**